### Problem 01 -- Word Counts

#### Part A

How many times are each of the following characters mentioned by name in the text of Little Women?
    • Jo, Beth, Meg, Amy

In [1]:
# First we get the text of Little Women.

!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt

--2016-09-25 03:26:21--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/women.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1053440 (1.0M) [text/plain]
Saving to: ‘women.txt’

women.txt           100%[=====================>]   1.00M  --.-KB/s   in 0.05s  

2016-09-25 03:26:21 (22.2 MB/s) - ‘women.txt’ saved [1053440/1053440]



In [2]:
""" For each name we read through the text, 
    keeping only the lines that have at least one occurrence of the name,
    split it into one word per line, sort the lines, 
    collapse lines counting unique words then sort again 
    and display only those that match the name.
    Here we count occurrences of 'Jo'. """

!cat women.txt | grep 'Jo' | grep -oE '\w{{2,}}' | sort | uniq -c | sort -rn | grep 'Jo'

   1355 Jo
    163 John
      6 Joanna
      4 Josephine
      3 Jove
      3 Josy
      2 Johnson
      1 Jouvin
      1 Josyphine
      1 Joneses
      1 Jones
      1 Joian
      1 Johnsonianly
      1 Johnsonian


In [3]:
#   And here are the counts for 'Beth', 'Meg' and 'Amy'.

!cat women.txt | grep 'Beth' | grep -oE '\w{{2,}}' | sort | uniq -c | sort -rn | grep 'Beth'
!cat women.txt | grep 'Meg' | grep -oE '\w{{2,}}' | sort | uniq -c | sort -rn | grep 'Meg'
!cat women.txt | grep 'Amy' | grep -oE '\w{{2,}}' | sort | uniq -c | sort -rn | grep 'Amy'

    459 Beth
      5 Bethy
      1 Beths
    683 Meg
      1 Megs
      1 Meggy
    645 Amy


So, in Little Women:
    Jo is mentioned by name 1355 times,
    Beth is mentioned by name 459 times,
    Meg is mentioned by name 683 times and
    Amy is mentioned by name 645 times.



### Problem 01 -- Word Counts

#### Part B

How many times do each of the characters Juliet and Romeo have speaking lines in Romeo and Juliet?

In [4]:
# First we get the text of Romeo & Juliet

!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt

--2016-09-25 03:26:56--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/projects/project-01/romeo.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.20.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.20.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 178983 (175K) [text/plain]
Saving to: ‘romeo.txt’

romeo.txt           100%[=====================>] 174.79K  --.-KB/s   in 0.02s  

2016-09-25 03:26:56 (7.78 MB/s) - ‘romeo.txt’ saved [178983/178983]



In [ ]:
""" We shall read through the text once for Romeo and once for Juliet
    and count the lines that contain 'Rom.' and those that contain 'Jul.'.
    That is because all speaking lines (no matter how long) begin with 
    'Rom.' for Romeo and 'Jul.' for Juliet. Even the first line of Romeo's 
    famous soliloquy (Rom.  To be or not to be...) begins with 'Rom.'. 
    So each instance of 'Rom.' denotes a speaking part for Romeo and each
    instance of 'Jul.' denotes a speaking part for Juliet.
    
    As such, the number of lines that include 'Rom.' is also the number of times
    that Romeo has a speaking part in the play.
"""

In [5]:
# Using a shell pipeline similar to the one deployed above, we count occurrences of 'Rom.' and 'Jul.'

# I have yet to figure out how to get Python/Unix shell to print the '.' in 'Rom.' or 'Jul.'
# I may have to declare the string 'Rom.' as literal or raw, but I'm not sure.
# In any event, it doesn't matter because whether or not we print 'Rom' or 'Rom.' 
# doesn't change the number of times Romeo or Juliet have speaking lines.

!cat romeo.txt | grep 'Rom.' | grep -oE '\w{{2,}}' | sort | uniq -c | sort -rn | head -5
!cat romeo.txt | grep 'Jul.' | grep -oE '\w{{2,}}' | sort | uniq -c | sort -rn | head -5

    163 Rom
    159 Romeo
     53 and
     49 to
     39 the
    117 Jul
     69 Juliet
     36 and
     28 thou
     27 is


So in Romeo & Juliet, 
    Romeo has speaking lines 163 times and
    Juliet has speaking lines 117 times.



### Problem 02 -- Capital Bikeshare

#### Part A

Which 10 Capital Bikeshare stations were the most popular departing stations in Q1 2016? Which 10 were
the most popular destination stations in Q1 2016?

In [ ]:
""" The tools in csvkit are perfect for this task. 

First we get data for Q1 2016 from Capital Bikeshare.
Then we slice & dice a few times with csvkit tools to determine:
    the top ten departing stations,
    the top ten destination stations 
    and the top ten bikes departing from and arriving at 
    the top stations for departures and arrivals.

"""

In [6]:
# First we get the data.

!wget https://s3.amazonaws.com/capitalbikeshare-data/2016-Q1-cabi-trip-history-data.zip

--2016-09-25 03:27:36--  https://s3.amazonaws.com/capitalbikeshare-data/2016-Q1-cabi-trip-history-data.zip
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.1.163
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.1.163|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 11048826 (11M) [application/zip]
Saving to: ‘2016-Q1-cabi-trip-history-data.zip’

2016-Q1-cabi-trip-h 100%[=====================>]  10.54M  27.7MB/s   in 0.4s   

2016-09-25 03:27:37 (27.7 MB/s) - ‘2016-Q1-cabi-trip-history-data.zip’ saved [11048826/11048826]



In [7]:
# Unzip it, call it q1.csv and see the variable names.

!unzip 2016-Q1-cabi-trip-history-data.zip

Archive:  2016-Q1-cabi-trip-history-data.zip
  inflating: 2016-Q1-Trips-History-Data.csv  


In [8]:
!mv 2016-Q1-Trips-History-Data.csv q1.csv

In [9]:
# Let's see what variables (data columns) we've got.

!csvcut -n q1.csv

  1: Duration (ms)
  2: Start date
  3: End date
  4: Start station number
  5: Start station
  6: End station number
  7: End station
  8: Bike number
  9: Member Type


In [10]:
# Check earliest and latest trips to see that the data span all of Q1 2016.
# Since we're not sorting here we're assuming Capital Bikeshare keeps it's data tidy.

!csvcut -c2,3 q1.csv | head -15 | csvlook
!csvcut -c2,3 q1.csv | tail -15 | csvlook

!csvcut -c5,7,8 q1.csv | head -5 | csvlook

|------------------+------------------|
|  Start date      | End date         |
|------------------+------------------|
|  3/31/2016 23:59 | 4/1/2016 0:04    |
|  3/31/2016 23:59 | 4/1/2016 0:08    |
|  3/31/2016 23:59 | 4/1/2016 0:08    |
|  3/31/2016 23:57 | 4/1/2016 0:09    |
|  3/31/2016 23:57 | 3/31/2016 23:59  |
|  3/31/2016 23:57 | 4/1/2016 0:13    |
|  3/31/2016 23:57 | 4/1/2016 0:06    |
|  3/31/2016 23:56 | 4/1/2016 0:00    |
|  3/31/2016 23:55 | 4/1/2016 0:01    |
|  3/31/2016 23:55 | 3/31/2016 23:59  |
|  3/31/2016 23:55 | 4/1/2016 0:03    |
|  3/31/2016 23:54 | 4/1/2016 0:01    |
|  3/31/2016 23:54 | 4/1/2016 0:22    |
|  3/31/2016 23:54 | 4/1/2016 0:23    |
|------------------+------------------|
|----------------+----------------|
|  1/1/2016 0:28 | 1/1/2016 0:34  |
|----------------+----------------|
|  1/1/2016 0:28 | 1/1/2016 0:35  |
|  1/1/2016 0:27 | 1/1/2016 0:32  |
|  1/1/2016 0:27 | 1/1/2016 0:34  |
|  1/1/2016 0:26 | 1/1/2016 0:40  |
|  1/1/2016 0:26 | 1/1/2016 

In [11]:
# Now we list the top ten stations ordered by number of trips departing in Q1 2016.

!csvcut -c5,4 q1.csv | sort | uniq -c | sort -rn | head -10 | csvlook

sort: write failed: standard output: Broken pipe
sort: write error
|--------------------------------------------------------------+--------|
|    13120 Columbus Circle / Union Station                     | 31623  |
|--------------------------------------------------------------+--------|
|     9560 Massachusetts Ave & Dupont Circle NW                | 31200  |
|     9388 Lincoln Memorial                                    | 31258  |
|     8138 Jefferson Dr & 14th St SW                           | 31247  |
|     7479 Thomas Circle                                       | 31241  |
|     7401 15th & P St NW                                      | 31201  |
|     6568 14th & V St NW                                      | 31101  |
|     6491 New Hampshire Ave & T St NW                         | 31229  |
|     5649 Eastern Market Metro / Pennsylvania Ave & 7th St SE | 31613  |
|     5514 17th & Corcoran St NW                               | 31214  |
|--------------------------------------------

In [12]:
# And just to check if the write errors are fatal, we do it again using only station number.
# No write errors this time with the same results, so the write errors are superficial.

!csvcut -c4 q1.csv | sort | uniq -c | sort -rn | head -10 | csvlook

|-----------------|
|    13120 31623  |
|-----------------|
|     9560 31200  |
|     9388 31258  |
|     8138 31247  |
|     7479 31241  |
|     7401 31201  |
|     6568 31101  |
|     6491 31229  |
|     5649 31613  |
|     5514 31214  |
|-----------------|


In [13]:
# Now we list the top ten stations ordered by number of trips finishing at each in Q1 2016.

!csvcut -c7 q1.csv | sort | uniq -c | sort -rn | head -10 | csvlook

sort: fflush failed: standard output: Broken pipe
sort: write error
|------------------------------------------------|
|    13880 Columbus Circle / Union Station       |
|------------------------------------------------|
|    11183 Massachusetts Ave & Dupont Circle NW  |
|     9419 Lincoln Memorial                      |
|     8975 Jefferson Dr & 14th St SW             |
|     8092 15th & P St NW                        |
|     7267 14th & V St NW                        |
|     6997 Thomas Circle                         |
|     6245 New Hampshire Ave & T St NW           |
|     5761 5th & K St NW                         |
|     5651 17th & Corcoran St NW                 |
|------------------------------------------------|


So, Columbus Circle / Union Station was by far the most popular departing 
and the most popular destination station in Q1 2016 (in terms of 
number of trips originating from and concluding at each station).
13120 trips originated and 13880 trips concluded at this station in Q1 2016.



### Problem 02 -- Capital Bikeshare

#### Part B

For the most popular departure station, which 10 bikes were used most in trips departing from there? Which
10 bikes were used most in trips ending at the most popular destination station?

In [14]:
# The most popular departure station in Q1 2016 was Union Station.

# In a single line pipeline, we generate a list of bikes that departed Union Station
# the most often sorted in descending order of departures.
# First it includes only the departing station and bike number columns, 
# keeps only Union Station records, sorts by bike number, 
# collapses multiple departures by each bike to one line each and prepends the number of departures,
# sorts in reverse order and prints the twenty bikes with most departures.

!csvcut -c5,8 q1.csv | csvgrep -c1 -m 'Columbus Circle / Union Station' | csvsort -c2 | uniq -c | sort -rn | head -20 | csvlook

|------------------------------------------+---------|
|       17 Columbus Circle / Union Station | W22227  |
|------------------------------------------+---------|
|       16 Columbus Circle / Union Station | W21867  |
|       16 Columbus Circle / Union Station | W21641  |
|       16 Columbus Circle / Union Station | W21538  |
|       16 Columbus Circle / Union Station | W21239  |
|       16 Columbus Circle / Union Station | W20540  |
|       16 Columbus Circle / Union Station | W00714  |
|       15 Columbus Circle / Union Station | W22080  |
|       15 Columbus Circle / Union Station | W21450  |
|       15 Columbus Circle / Union Station | W21076  |
|       15 Columbus Circle / Union Station | W00777  |
|       15 Columbus Circle / Union Station | W00288  |
|       14 Columbus Circle / Union Station | W22099  |
|       14 Columbus Circle / Union Station | W22079  |
|       14 Columbus Circle / Union Station | W21623  |
|       14 Columbus Circle / Union Station | W20425  |
|       14

The ten bikes used the most times in Q1 2016 to leave the most popular departure station (Union Station) are listed in descending order of departures in the table above. Bike number W22227 left 17 times, more than any other bike. Six bikes left 16 times each (all members of the top ten) and five bikes left 15 times (so any three of these round out the last three slots in the list of top ten departing bikes from the most popular departing station).

In [15]:
# This is the same single line pipeline as above but for Union Station as destination station.

!csvcut -c7,8 q1.csv | csvgrep -c1 -m 'Columbus Circle / Union Station' | csvsort -c2 | uniq -c | sort -rn | head -20 | csvlook

sort: write failed: standard output: Broken pipe
sort: write error
|------------------------------------------+---------|
|       18 Columbus Circle / Union Station | W00485  |
|------------------------------------------+---------|
|       17 Columbus Circle / Union Station | W22227  |
|       16 Columbus Circle / Union Station | W22099  |
|       16 Columbus Circle / Union Station | W22080  |
|       16 Columbus Circle / Union Station | W21239  |
|       16 Columbus Circle / Union Station | W21076  |
|       16 Columbus Circle / Union Station | W20425  |
|       16 Columbus Circle / Union Station | W00714  |
|       15 Columbus Circle / Union Station | W21997  |
|       15 Columbus Circle / Union Station | W21867  |
|       15 Columbus Circle / Union Station | W21641  |
|       15 Columbus Circle / Union Station | W21538  |
|       15 Columbus Circle / Union Station | W21450  |
|       15 Columbus Circle / Union Station | W20540  |
|       15 Columbus Circle / Union Station | W01439  

The ten bikes used the most times in Q1 2016 to come to the most popular arrival station (Union Station) are listed in descending order of arrivals in the table above. Bike number W00485 arrived 18 times, more than any other bike. Bike number W22227--the bike used the most to leave Union Station in Q1--was a close second, arriving 17 times. Six bikes ended their trips 16 times each (all members of the top ten) and seven bikes arrived 15 times (so any two of these round out the last two slots in the list of top ten bikes arriving at the most popular arrival station in Q1 2016).



### Problem 03 -- Filters

I will be working through this section this weekend as I'm not as familiar with filters in Python as I need to be. 
